lib/segment/src/index/query_optimization/condition_converter.rs

```diff
@@ -56,8 +59,63 @@ pub fn condition_converter<'a>(
                 .collect();
             Box::new(move |point_id| segment_ids.contains(&point_id))
         }
+        Condition::Nested(nested) => {
+            // Select indexes for nested fields. Trim nested part from key, so
+            // that nested condition can address fields without nested part.
+
+            // Example:
+            // Index for field `nested.field` will be stored under key `nested.field`
+            // And we have a query:
+            // {
+            //   "nested": {
+            //     "path": "nested",
+            //     "filter": {
+            //         ...
+            //         "match": {"key": "field", "value": "value"}
+            //     }
+            //   }
+
+            // In this case we want to use `nested.field`, but we only have `field` in query.
+            // Therefore we need to trim `nested` part from key. So that query executor
+            // can address proper index for nested field.
+
+            let nester_indexes: HashMap<_, _> = field_indexes
+                .iter()
+                .filter_map(|(key, indexes)| {
+                    if key.starts_with(&nested.array_key()) {
+                        // Trim nested part from key
+                        let nested_key = key[(nested.array_key().len() + 1)..].to_string();
+                        Some((nested_key, indexes))
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            Box::new(move |point_id| {
+                payload_provider.with_payload(point_id, |payload| {
+                    let field_values = payload.get_value(&nested.array_key()).values();
+
+                    for value in field_values {
+                        if let Value::Object(object) = value {
+                            let get_payload = || OwnedPayloadRef::from(object);
+                            if check_payload(
+                                get_payload,
+                                |_| None,
+                                &nested.nested.filter,
+                                point_id,
+                                &nester_indexes,
+                            ) {
+                                // If at least one nested object matches, return true
+                                return true;
+                            }
+                        }
+                    }
+                    false
```

Comment:
We can do some iterator coolness here:

```suggestion
                    payload
                        .get_value(&nested.array_key())
                        .values()
                        .iter()
                        .filter_map(|value| match value {
                            Value::Object(object) => Some(object),
                            _ => None,
                        })
                        // If at least one nested object matches, return true
                        .any(|object| {
                            check_payload(
                                || OwnedPayloadRef::from(object),
                                |_| None,
                                &nested.nested.filter,
                                point_id,
                                &nester_indexes,
                            )
                        })
```